### ISSUE TITLE
[2020]: yolov4

### ISSUE THREAD

**Issue by TakeruEndo** [TakeruEndo/paper-survey#22]
## 概要
大規模なデータセットに適応させるため以下の機能を追加した。
- Weighted-Residual-Connections (WRC)
- Cross-Stage-Partial-connections (CSP)
- Cross mini-Batch
- Normalization (CmBN)
- Self-adversarial-training (SAT)
- Mish-activation

## 導入
車両の衝突事故などではリアルタイムに物体検出しないといけない。

### 本論文の貢献
- 1080 Ti or 2080 Ti GPUで動作する効率的で強力な物体検出モデルの作成
- Bag-of FreebiesとBag-of-Specialsの検証
- CBN, PAN, SAM......

<img width="524" alt="スクリーンショット 2020-03-30 16 51 22" src="https://user-images.githubusercontent.com/40351074/81264335-2cc9f980-907c-11ea-9cb9-5c388cd0534c.png">

### Bag-of Freebies
オフラインでの学習の利点を生かし、訓練戦略を変更するだけの方法や、訓練コストを増加させるだけの方法  。その例はデータ拡張である。データ拡張の目的は入力画像のばらつきを増加させて、設計された物体検出モデルが異なる環境から得られた画像に関してより高いロバスト性を持たせることにある。  
フォトメトリック歪みと幾何学的歪みは物体検出において確実に利益をもたらす。
#### 有用なデータ拡張の例
- DropOut
- DropConnect
- DropBlock
- MixUp
- CutMix
- style transfer GAN によるデータの水増し

#### データ不均衡の解消
- [hard negative example mining](https://www.researchgate.net/publication/3192800_Example_Based_Learning_for_View-Based_Human_Face_Detection)
- [r online hard example mining](https://arxiv.org/abs/1604.03540)
しかし、上記の手法はtwo_stage検出器には適用できない。one_stage検出器ではfocal_lossが使われる。

#### 物体検出における目的関数
- DIoU loss
- CIOU loss
- GIOU loss

## Bag of specials
- enlarging receptive field
- attention
- strengthening feature integration capability

#### モジュール
- SPPモジュール
- ASPP

#### Attention
- Squeeze-and-Excitation (SE) 
- Spatial Attention Module (SAM)

#### 後処理
- NMS

## 方策
ニューラルネットワークの動作速度の高速化のために、生産システムでのリアルタイム計算と並列計算の最適化をする。

## アーキテクチャーの選定
- CSPDarknet53の上にSPPブロックを追加
- FPNの代わりにPANetを使用
- CSPDarknet53バックボーン、SPP追加モジュール、PANetパスアグリゲーションネック、YOLOv3(アンカーベース)ヘッド

### 活性化関数
- ReLU, leaky-ReLU, parametric-ReLU, ReLU6, SELU, Swish, or Mish
### Bounding boxの回帰損失
- MSE, IoU, GIoU, CIoU, DIoU
### データ拡張
- CutOut, MixUp, CutMix
### 正規化
- DropOut, DropPath, Spatial DropOut, DropBlock
### Normalization of the network activations by their
mean and variance: 
- Batch Normalization (BN) [32],
- Cross-GPU Batch Normalization (CGBN or SyncBN)
- Filter Response Normalization (FRN)
- Cross-Iteration Batch Normalization (CBN) [89]
### Skip-connections: 
- Residual connections, Weighted
residual connections
- Multi-input weighted residual connections
- Cross stage partial connections (CSP)

### Self-Adversarial Training (SAT)
第1段階では、ニューラルネットワークはネットワークの重みの代わりに元の画像を変更する。このようにして、ニューラルネットワークは、元の画像を変更して、画像上に目的の物体が存在しないという欺瞞を作り出すために、自分自身に対して敵対的な攻撃を実行できる。第2段階では、ニューラルネットワークは、この変更された画像上の物体を通常の方法で検出するように訓練されます。

## YOLOv4
- Backbone: CSPDarknet53 [81]
- Neck: SPP [25], PAN [49]
- Head: YOLOv3 [63]

### YOLO v4の詳細
- Bag of Freebies (BoF) for backbone:   
`CutMix`, `Mosaic data augmentation`, `DropBlock regularization`, `Class label smoothing`
- Bag of Specials (BoS) for backbone:   
Mish activation, Cross-stage partial connections (CSP), Multiinput weighted residual connections (MiWRC)
- Bag of Freebies (BoF) for detector:  
CIoU-loss, CmBN, DropBlock regularization, Mosaic data augmentation, Self-Adversarial Training, Eliminate grid sensitivity, Using multiple anchors for a single ground truth, Cosine annealing scheduler [52], Optimal hyperparameters, Random training shapes
• Bag of Specials (BoS) for detector:  
Mish activation, SPP-block, SAM-block, PAN path-aggregation block,
DIoU-NMS

<img width="524" alt="スクリーンショット 2020-03-30 16 51 22" src="https://user-images.githubusercontent.com/40351074/81264257-0b690d80-907c-11ea-9a53-3e469076ea36.png">

<img width="524" alt="スクリーンショット 2020-03-30 16 51 22" src="https://user-images.githubusercontent.com/40351074/81264440-5c790180-907c-11ea-8f36-8413bd9ab0c7.png">

## 論文情報・リンク
- https://github.com/AlexeyAB/darknet

* [Alexey Bochkovskiy, Chien-Yao Wang, Hong-Yuan Mark Liao，"Optimal Speed and Accuracy of Object Detection，2020](https://arxiv.org/abs/2004.10934)
